src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,448 @@
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const MY_KEY: i32 = 1337;
+    ///
+    /// let mut shared_memory = SharedMemory::<MyData>::new(
+    ///     MY_KEY,
+    ///     None,
+    ///     ShmatFlag::empty(),
+    ///     Mode::empty(),
+    /// )?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    pub fn new(
+        shmid: c_int,
+        shmaddr: Option<c_void>,
+        shmat_flag: ShmatFlag,
+        mode: Mode,
+    ) -> Result<Self> {
+        unsafe {
+            Ok(Self {
+                id: shmid,
+                shm: ManuallyDrop::new(Box::from_raw(Self::shmat(
+                    shmid, shmaddr, shmat_flag, mode,
+                )?)),
+            })
+        }
+    }
+
+    /// Creates and returns a new, or returns an existing, System V shared memory
+    /// segment identifier.
+    ///
+    /// For more information, see [`shmget(2)`].
+    ///
+    /// # Example
+    ///
+    /// ## Creating a shared memory zone
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const MY_KEY: i32 = 1337;
+    ///
+    /// let id = SharedMemory::<MyData>::shmget(
+    ///     MY_KEY,
+    ///     ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL,
+    ///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    /// )?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    /// [`shmget(2)`]: https://man7.org/linux/man-pages/man2/shmget.2.html
+    pub fn shmget(
+        key: key_t,
+        shmget_flag: ShmgetFlag,
+        mode: Mode,
+    ) -> Result<i32> {
+        let size = std::mem::size_of::<T>();
+        let flags = mode.bits() as i32 | shmget_flag.bits();
+        Errno::result(unsafe { libc::shmget(key, size, flags) })
+    }
+
+    /// Performs control operation specified by `cmd` on the System V shared
+    /// memory segment given by `shmid`.
+    ///
+    /// For more information, see [`shmctl(2)`].
+    ///
+    /// # Example
+    ///
+    /// ## Deleting a shared memory zone
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const ID: i32 = 1337;
+    ///
+    /// let mut shared_memory = SharedMemory::<MyData>::new(
+    ///     ID,
+    ///     None,
+    ///     ShmatFlag::empty(),
+    ///     Mode::empty(),
+    /// )?;
+    ///
+    /// let _ = shared_memory.shmctl(ShmctlFlag::IPC_RMID, None, Mode::empty())?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    /// [`shmctl(2)`]: https://man7.org/linux/man-pages/man2/shmctl.2.html
+    pub fn shmctl(
+        &self,
+        shmctl_flag: ShmctlFlag,
+        buf: Option<shmid_ds>,
```

Comment:
What about passing it by reference:

```suggestion
        buf: Option<&shmid_ds>,
```

Reply:
Yes, it's even transformed to a pointer after. Fixed